### PR TITLE
Add Bootstrap 5 support

### LINKED ISSUE #777: Any plans for BS5 support?
Since Bootstrap v5 is about to go beta (last breaking changes implemented), are there plan to make bootbox BS5 friendly?

### PR DESCRIPTION
**Pull request by jstark518**
- Fixes #777 - Builds on PR791: Add support for bootstrap 5 while keeping support for 4
- Changed test to use Chrome
- Changed test to look for .btn-close when using BS5, .close when using BS4.
- Added another test config, test both BS4 and BS5.
- Change to @popperjs/core from popper, see PR746.

### CODE REVIEW

File: bootbox.js [hunk 1]
```diff
@@ -789,7 +789,9 @@
         });
 
         // safe to set a select's value as per a normal input
-        input.val(options.value);
```
Comment (jstark518) [inline bootbox.js:792]: #798 This was causing problems with the existing test

**Comment by tarlepp** [comment]
Hmm, if this is for BS5 and BS4 then we should - imho - have test matrix for each of those - to make sure that all works as expected.

File: bootbox.js (hunk 1 above)
Comment (jstark518) [reply bootbox.js:792]: http://jsfiddle.net/osuzjbxg/1

File: package.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
     "test": "tests"
   },
   "scripts": {
-    "test": "./node_modules/.bin/karma start"
+    "test": "./node_modules/.bin/karma start karma.conf.js && ./node_modules/.bin/karma start karma-bs4.conf.js"
```
Comment (jstark518) [inline package.json:9]: @tarlepp

**Comment by rajpt** [comment]
Good Stuff. Just waiting to get this merged and have bootstrap 5 support.

**Comment by Yohn** [comment]
Thanks for this PR!

One thing I found is when modal-header is not there, the X / close button still appears but within the modal-body, and is on the left side. Looks out of place. I'll let you know if I find anything else popping up while I play with it for a bit.

**Comment by jstark518** [comment]
> Thanks for this PR!
> 
> One thing I found is when modal-header is not there, the X / close button still appears but within the modal-body, and is on the left side. Looks out of place. I'll let you know if I find anything else popping up while I play with it for a bit.

Thank you, I'll fix that Monday.

**Comment by jstark518** [comment]
> Thanks for this PR!
> 
> One thing I found is when modal-header is not there, the X / close button still appears but within the modal-body, and is on the left side. Looks out of place. I'll let you know if I find anything else popping up while I play with it for a bit.

@Yohn Fixed

**Comment by bobrundle** [comment]
Would like to see this PR merged.  I have a site that uses bootbox heavily and I am updating it to bootstrap 5.  bootbox and a few other packages are holding me back.  If there was a replacement to bootbox I would use it.  I spent a day trying to hammer SweetAlert2 in only to give up.  Nothing out there that I can find has the dialog functionality.

**Comment by jstark518** [comment]
> Would like to see this PR merged. I have a site that uses bootbox heavily and I am updating it to bootstrap 5. bootbox and a few other packages are holding me back. If there was a replacement to bootbox I would use it. I spent a day trying to hammer SweetAlert2 in only to give up. Nothing out there that I can find has the dialog functionality.

@bobrundle I wouldn't hold out hope it gets merged, do you use scss?

**Comment by bobrundle** [comment]
@jstark518 I would not be so pessimistic.  Your BS5 support will be merged eventually.  Things move slow here as they often when free labor is at work.  BS5 support is a major improvement and needs to be carefully evaluated.

The site I am working on does not use scss but I have integrated your changes into it as I await the official bootbox BS5 support.

**Comment by jstark518** [comment]
> Can't wait for Bootstrap 5 support, this is my workaround written in Sass based on Bootstrap 5 .btn-close, changed the color to transparant to prevent a double cross icon. Hope this temporary fix is helpfull:
> 
> ```
> .bootbox-close-button {
>     box-sizing: content-box;
>     width: $btn-close-width;
>     height: $btn-close-height;
>     padding: $btn-close-padding-y $btn-close-padding-x;
>     // color: $btn-close-color;
>     color: transparent;
>     background: transparent escape-svg($btn-close-bg) center / $btn-close-width auto no-repeat; // include transparent for button elements
>     border: 0; // for button elements
>     @include border-radius();
>     opacity: $btn-close-opacity;
> 
>     // Override <a>'s hover style
>     &:hover {
>         // color: $btn-close-color;
>         color: transparent;
>         text-decoration: none;
>         opacity: $btn-close-hover-opacity;
>     }
> 
>     &:focus {
>         outline: 0;
>         box-shadow: $btn-close-focus-shadow;
>         opacity: $btn-close-focus-opacity;
>     }
> 
>     &:disabled,
>     &.disabled {
>         pointer-events: none;
>         user-select: none;
>         opacity: $btn-close-disabled-opacity;
>     }
> }
> ```

I did the same exact thing

**Comment by erobertson42** [comment]
> Would like to see this PR merged. I have a site that uses bootbox heavily and I am updating it to bootstrap 5. bootbox and a few other packages are holding me back. If there was a replacement to bootbox I would use it. I spent a day trying to hammer SweetAlert2 in only to give up. Nothing out there that I can find has the dialog functionality.

https://github.com/erobertson42/bootstrap-modbox

Yes, this is my library.  It is vanilla JavaScript for Bootstrap 5, no dependencies, and uses standard Bootstrap classes/styling with no custom CSS.  So if you're already using a theme or your own Bootstrap Sass modifications, they should automatically "just work".

I was on the fence about replying with this, as I wasn't sure if it was bad form or not, though that wasn't my intention.  Mods, if you disapprove of me posting this here, I understand if you remove it - apologies and no hard feelings.  I'm not really out for advertising my library, I've never even posted about it anywhere before this, just trying to help with a possible alternative.

I also wanted to mention that while the library has no relation to bootbox, I did model the docs site layout after bootbox's (imitation -> flattery 🙂 ).

**Comment by erobertson42** [comment]
> @erobertson42 I won't pretend it's the best example, but this uses Jekyll: https://github.com/tiesont/bootboxdocs

Great, thank you. I'll take a look when I get a chance.

**Comment by picov** [comment]
Any news about BS5 support in an official release?

**Comment by tiesont** [comment]
> Any news about BS5 support in an official release?

Depends on what you mean by "BS5 support". _Properly_ supporting Bootstrap 5 requires rewriting Bootbox to eliminate our jQuery dependencies, adjust for changes in the JavaScript functions, and figure out how to keep a fair amount of backward compatibility. 

None of those are trivial tasks, and I haven't had a ton of free time lately. The above changes also necessitate a major version bump, which I'm not entirely sure I want to commit to at this point. @erobertson42 's project looks pretty promising, so if you need something _right now_, I'd suggest taking a look at https://github.com/erobertson42/bootstrap-modbox

Probably not the answer you were looking for, but it's all I have for you (and others) at the moment.

**Comment by tarlepp** [comment]
And really if _we're_ going to add that BS5 support - imho - we should just create new major version, just because we could just drop all the old things with that - there haven't been so much changes in years in this repository so I think that would be the "proper" way to add BS5 support.

Although we haven't hear anything from @makeusabrew for a while (we've tried to contact him, but no luck yet - we hope that he's ok), but because we haven't hear anything from him a long time - I think that someone should make fork of this repository and continue working within that with new features.

And yes there is couple of those forks already - but I think that we should choose which one to start to use - and recommend that here too.

And I really hope that new fork repository maintainers (see the plurar format) could provide their time to maintain and development this great library.

**Comment by tiesont** [comment]
> And really if _we're_ going to add that BS5 support - imho - we should just create new major version, just because we could just drop all the old things with that - there haven't been so much changes in years in this repository so I think that would be the "proper" way to add BS5 support.
> 
> Although we haven't hear anything from @makeusabrew for a while (we've tried to contact him, but no luck yet - we hope that he's ok), but because we haven't hear anything from him a long time - I think that someone should make fork of this repository and continue working within that with new features.
> 
> And yes there is couple of those forks already - but I think that we should choose which one to start to use - and recommend that here too.
> 
> And I really hope that new fork repository maintainers (see the plurar format) could provide their time to maintain and development this great library.

I did create https://github.com/BootboxJS/bootbox at one point - do you think it makes more sense to just work from that project, since we'd have full control of all settings, or do enough work *here* to make it fully BS5 compatible and then basically halt development?

I do still use Bootbox on a daily basis, but I haven't been able to upgrade those projects to BS5 yet (heck, one is stuck on 3.4 until I can convince the bossmen that it _really_ needs to be updated), so the current Bootbox works just fine for me.

I really don't want to abandon this repo, but without @makeusabrew popping in long enough to transfer ownership, it doesn't really make sense to keep working here long term - we can't add any more maintainers or pretty much anything else that requires admin privileges...

**Comment by jstark518** [comment]
According to Bootstrap'd doc, jQuery works fine.

[](https://getbootstrap.com/docs/5.0/getting-started/javascript/#still-want-to-use-jquery-its-possible)Still want to use jQuery? It’s possible!
Bootstrap 5 is designed to be used without jQuery, but it’s still possible to use our components with jQuery. If Bootstrap detects jQuery in the window object it’ll add all of our components in jQuery’s plugin system; this means you’ll be able to do $('[data-bs-toggle="tooltip"]').tooltip() to enable tooltips. The same goes for our other components.

**Comment by tiesont** [comment]
> According to Bootstrap'd doc, jQuery works fine.
> 
> Still want to use jQuery? It’s possible! Bootstrap 5 is designed to be used without jQuery, but it’s still possible to use our components with jQuery. If Bootstrap detects jQuery in the window object it’ll add all of our components in jQuery’s plugin system; this means you’ll be able to do $('[data-bs-toggle="tooltip"]').tooltip() to enable tooltips. The same goes for our other components.

I'm aware of that, but it's also possible to use Bootstrap 5 _without_ jQuery. We use jQuery internally to build the modal (among other things), so for _real_ Bootstrap 5 compatibility, we need to assume jQuery _won't_ be used.

**Comment by tiesont** [comment]
@jstark518 I don't think I'm going to merge this pull-request. I did fix some of the issues you noted, in the master branch, but I think I want to take a slightly different approach to maintaining Bootstrap 4.6 compatibility. It's up to you at this point whether you want to leave the pull-request open or not.

I've deleted and then recreated the v5.x branch, to serve as a snapshot of the current Bootbox version (5.5.2), and am going to spend some time this week working on a basic v6 version. Not going to drop jQuery at this point, but I have a few relatively minor updates I can make to generate the correct markup for both Bootstrap 4 and Bootstrap 5, and add a few features that are in a customized version I use at work.

**Comment by jstark518** [comment]
> @jstark518 I don't think I'm going to merge this pull-request. I did fix some of the issues you noted, in the master branch, but I think I want to take a slightly different approach to maintaining Bootstrap 4.6 compatibility. It's up to you at this point whether you want to leave the pull-request open or not.
> 
> I've deleted and then recreated the v5.x branch, to serve as a snapshot of the current Bootbox version (5.5.2), and am going to spend some time this week working on a basic v6 version. Not going to drop jQuery at this point, but I have a few relatively minor updates I can make to generate the correct markup for both Bootstrap 4 and Bootstrap 5, and add a few features that are in a customized version I use at work.

Looking forward to it. 

I like the idea of keeping jQuery.

**Comment by jstark518** [comment]
Need to delete the repo to make a new fork

**Pull request closed without merging**